### PR TITLE
Java: data flow for `java.util.Objects`

### LINKED ISSUE #3895: Java: Model taint for `java.util.Objects`
**Description of the issue**
It appears QL is currently not tainting return values of the [`java.util.Objects`](https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/util/Objects.html) methods. Would it be possible to add this functionality?
Methods:
- `requireNonNull(T)`
- `requireNonNull​(T, String)`
- `requireNonNull​(T, Supplier<String>)`
- `requireNonNullElse​(T, T)` (taint both arguments)
- `requireNonNullElseGet​(T, Supplier<T>)`
- ? `toString(Object)`
- `toString(Object, String)` (at least taint the `String` argument)

### PR DESCRIPTION
**Pull request by aibaars**
Model taint propagation to return values  for methods of `java.util.Objects` [`java.util.Objects`](https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/util/Objects.html) 

Methods:
- `requireNonNull(T)`
- `requireNonNull​(T, String)`
- `requireNonNull​(T, Supplier<String>)`
- `requireNonNullElse​(T, T)` (taint both arguments)
- `requireNonNullElseGet​(T, Supplier<T>)`
- ~~`toString(Object)`~~
- `toString(Object, String)` (taint only second argument)

Fixes #3895

### CODE REVIEW

**Comment by aschackmull** [comment]
Some of these are actually value-preserving steps and should be included in `simpleLocalFlowStep` in `DataFlowUtil.qll` instead of as taint steps. The taint step from a `Supplier<T>` is not going to work on its own, but I don't think it's an important flow step, so whether or not we include it won't really matter. I'm not sure whether we should include the general `toString` steps as taint steps. This is an area where we aren't very consistent - we've generally only included `obj.toString()` as a taint step for overrides on certain types, and we're currently not including the step when the static target is `Object.toString()`. However, we do include `"" + obj` as a taint step (likely by coincidence), which is unfortunately somewhat inconsistent.

**Comment by aschackmull** [comment]
Also, it looks like the `ObjectsTest.java` file is missing.

**Comment by aibaars** [comment]
I left `Objects.toString(..., tainted)` as a flow step.

File: java/ql/test/library-tests/dataflow/local-flow/ObjectsTest.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import java.util.Objects;
+
+class ObjectsTest {
+	public static void taintSteps() {
```
Comment (aschackmull) [inline java/ql/test/library-tests/dataflow/local-flow/ObjectsTest.java:4]:
```suggestion
	public static void valueSteps() {
```

File: java/ql/test/library-tests/dataflow/local-flow/flow.ql [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import java
+import semmle.code.java.dataflow.TaintTracking
+
+class Conf extends TaintTracking::Configuration {
```
Comment (aschackmull) [inline java/ql/test/library-tests/dataflow/local-flow/flow.ql:4]:
```suggestion
class Conf extends DataFlow::Configuration {
```

File: java/ql/test/library-tests/dataflow/local-flow/flow.ql [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import java
+import semmle.code.java.dataflow.TaintTracking
```
Comment (aschackmull) [inline java/ql/test/library-tests/dataflow/local-flow/flow.ql:2]:
```suggestion
import semmle.code.java.dataflow.DataFlow
```

**Pull request merged**